File: templatemethod/python/mytemplatemethod.py
class Pizza:
    def prepare(self):
        self.make_dough()
        self.add_sauce()
        self.add_toppings()
        self.bake()

    def make_dough(self):
        print("Making dough")

    def add_sauce(self):
        print("Adding sauce")

    def add_toppings(self):
        raise NotImplementedError("Subclasses must implement add_toppings method")

    def bake(self):
        print("Baking pizza")
        
    def slice(self):
        print("Slicing pizza")
        
    def box(self):
        print("Boxing pizza")

class PepperoniPizza(Pizza):
    def add_toppings(self):
        print("Adding pepperoni toppings")

class CheesePizza(Pizza):
    def add_toppings(self):
        print("Adding cheese toppings")

def main():
    pepperoni_pizza = PepperoniPizza()
    pepperoni_pizza.prepare()
    pepperoni_pizza.slice()
    pepperoni_pizza.box()

    cheese_pizza = CheesePizza()
    cheese_pizza.prepare()
    cheese_pizza.slice()
    cheese_pizza.box()

if __name__ == "__main__":
    main()

File: templatemethod/swift/mytemplatemethod.swift
import Foundation

// Abstract Base Class
class Pizza {
    // The Template Method: defines the skeleton of the algorithm
    final func makePizza() {
        prepareDough()
        addSauce()
        addToppings() // Custom step
        bake()
        box()
    }

    private func prepareDough() {
        print("Kneading and stretching the dough.")
    }

    private func addSauce() {
        print("Adding organic tomato sauce.")
    }

    private func bake() {
        print("Baking for 15 minutes at 450°F.")
    }

    private func box() {
        print("Placing pizza in a recycled cardboard box.")
    }

    // Hook: Subclasses must implement this to provide specific toppings
    func addToppings() {
        fatalError("Subclasses must implement addToppings()")
    }
}

// Concrete Implementation 1
class CheesePizza: Pizza {
    override func addToppings() {
        print("Adding generous layers of Mozzarella and Parmesan.")
    }
}

// Concrete Implementation 2
class PepperoniPizza: Pizza {
    override func addToppings() {
        print("Adding spicy pepperoni slices and Italian herbs.")
    }
}

// Main Execution
print("--- Ordering a Cheese Pizza ---")
let cheesePizza = CheesePizza()
cheesePizza.makePizza()

print("\n--- Ordering a Pepperoni Pizza ---")
let pepperoniPizza = PepperoniPizza()
pepperoniPizza.makePizza()
